crates/bevy_reflect/src/path/parse.rs

```diff
@@ -0,0 +1,174 @@
+use std::{fmt, num::ParseIntError};
+
+use thiserror::Error;
+
+use super::{Access, ReflectPathError};
+
+#[derive(Debug, PartialEq, Eq, Error)]
+#[error(transparent)]
+pub struct ParseError<'a>(Error<'a>);
+
+/// A parse error for a path string.
+#[derive(Debug, PartialEq, Eq, Error)]
+enum Error<'a> {
+    #[error("expected an identifier, but reached end of path string")]
+    NoIdent,
+
+    #[error("expected an identifier, got '{0}' instead")]
+    ExpectedIdent(Token<'a>),
+
+    #[error("failed to parse index as integer")]
+    IndexParse(#[from] ParseIntError),
+
+    #[error("A '[' wasn't closed, reached end of path string before finding a ']'")]
+    Unclosed,
+
+    #[error("A '[' wasn't closed properly, got '{0}' instead")]
+    BadClose(Token<'a>),
+
+    #[error("A ']' was found before an opening '['")]
+    CloseBeforeOpen,
+}
+
+pub(super) struct PathParser<'a> {
+    path: &'a str,
+    offset: usize,
+}
+impl<'a> PathParser<'a> {
+    pub(super) fn new(path: &'a str) -> Self {
+        PathParser { path, offset: 0 }
+    }
+
+    fn next_token(&mut self) -> Option<Token<'a>> {
+        // TODO(perf): using get_unchecked SAFETY: self.offset is never larger than self.path
+        // & self.offset always splits on UTF-8 sequence boundaries
+        let input = &self.path[self.offset..];
```

Comment:
I've kept this as a question for reviewers: should we use the `unsafe` version instead?

Reply:
What about keeping two fields, one for the whole path and one for the remaining input? You can get the updated remaining input by using either `.as_str()` on the `Chars` after calling `.next()` on it or as the second element of the tuple returned by `split_once`, both of which require no bound/utf8 checks. You can also get back the offset by subtracting the lengths of the two strings so you don't have to store it anymore.

Reply:
The initial draft actually did something similar: `self.path = self.path[ident.len()..]` instead of `self.offset += ident.len()`.

As you point out, I can't get back the original path if this. Keeping both works though. I really like your suggestion

Reply:
There is still a bound check with `split_at` (can't use `split_once` since it "eats" the `TOKEN` and we need to keep it) But I think it's better than have the bound check in the general case. I'll inspect the assembly output

Reply:
I did try using a `chars: Chars` field over `offset`, and I found it both more difficult to read and less performant, so I will keep it as is for now.